internal/venice-common/src/main/java/com/linkedin/venice/utils/SharedObjectFactory.java

```diff
@@ -24,17 +32,22 @@ public class SharedObjectFactory<T> {
    * @return A shared object
    */
   public T get(String identifier, Supplier<T> constructor, Consumer<T> destroyer) {
-    if (!objectMap.containsKey(identifier)) {
-      T object = constructor.get();
-      objectMap.put(identifier, new ReferenceCounted<>(object, obj -> {
-        destroyer.accept(obj);
-        objectMap.remove(identifier);
-      }));
-      return object;
-    } else {
-      ReferenceCounted<T> referenceCounted = objectMap.get(identifier);
-      referenceCounted.retain();
+    getLock(identifier).lock();
+    try {
+      final AtomicBoolean newObjectCreated = new AtomicBoolean(false);
+      ReferenceCounted<T> referenceCounted = OBJECT_MAP.computeIfAbsent(identifier, id -> {
+        newObjectCreated.set(true);
+        T object = constructor.get();
+        return new ReferenceCounted<>(object, destroyer);
+      });
+
+      if (!newObjectCreated.get()) {
+        referenceCounted.retain();
+      }
+
       return referenceCounted.get();
+    } finally {
+      getLock(identifier).unlock();
```

Comment:
Why lookup the lock twice? Can we look it up once at the beginning of the function, store it in a variable, and keep using it?